lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,422 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from __future__ import print_function
+import contextlib
+import multiprocessing
+import os
+import signal
+import sys
+import time
+
+try:
+    import termios
+except ImportError:
+    termios = None
+
+import pytest
+
+import llnl.util.tty.log
+from llnl.util.lang import uniq
+from llnl.util.tty.log import log_output
+from llnl.util.tty.pty import PseudoShell
+
+from spack.util.executable import which
+
+
+@contextlib.contextmanager
+def nullcontext():
+    yield
+
+
+def test_log_python_output_with_echo(capfd, tmpdir):
+    with tmpdir.as_cwd():
+        with log_output('foo.txt', echo=True):
+            print('logged')
+
+        # foo.txt has output
+        with open('foo.txt') as f:
+            assert f.read() == 'logged\n'
+
+        # output is also echoed.
+        assert capfd.readouterr()[0] == 'logged\n'
+
+
+def test_log_python_output_without_echo(capfd, tmpdir):
+    with tmpdir.as_cwd():
+        with log_output('foo.txt'):
+            print('logged')
+
+        # foo.txt has output
+        with open('foo.txt') as f:
+            assert f.read() == 'logged\n'
+
+        # nothing on stdout or stderr
+        assert capfd.readouterr()[0] == ''
+
+
+def test_log_python_output_and_echo_output(capfd, tmpdir):
+    with tmpdir.as_cwd():
+        # echo two lines
+        with log_output('foo.txt') as logger:
+            with logger.force_echo():
+                print('force echo')
+            print('logged')
+
+        # log file contains everything
+        with open('foo.txt') as f:
+            assert f.read() == 'force echo\nlogged\n'
+
+        # only force-echo'd stuff is in output
+        assert capfd.readouterr()[0] == 'force echo\n'
+
+
+@pytest.mark.skipif(not which('echo'), reason="needs echo command")
+def test_log_subproc_and_echo_output(capfd, tmpdir):
+    echo = which('echo')
+
+    with tmpdir.as_cwd():
+        # ensure that a subprocess echoing output is also logged
+        with log_output('foo.txt') as logger:
+            with logger.force_echo():
+                echo('echo')
+            print('logged')
+
+        with open('foo.txt') as f:
+            assert f.read() == 'logged\n'
```

Comment:
(question) why does this not say `'echo\nlogged\n'`?

Reply:
Yeah good question.  I had just left this in place, as these tests have been present since the original version of this code.  But it looks like this is because of interference like you're describing.  It only seems to happen when using a subprocess (`echo`) so I split this specific test into two -- one to test the logged output (with `capfd.disabled()`) and one to test the echoed output (with `capfd`).  This does not seem to be necessary for the ones that just call `print()`.